### PR TITLE
Import container disk image from registry

### PR DESCRIPTION
**Pull request by annastopel**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
CDI import from registry supports the same container image format as  kubevirt containerdisks
```

### CODE REVIEW

File: doc/image-from-registry.md [hunk 1]
```diff
@@ -2,47 +2,50 @@
 The purpose of this document is to show how to create registry image containing a Virtual Machine image that can be imported into a PV.
 
 ## Prerequisites
-The VM disk image should be called 'disk.img' and placed in the root folder.
+Import from registry should be able to consume the same container images as containerDisk (https://github.com/kubevirt/kubevirt/blob/master/docs/container-register-disks.md).
+Thus the VM disk image file to be consumed must be located under /disk directory in the container image. The file can be in any of the supported formats : qcow2, raw, archived. There is no any special naming contraints for the VM disk file.
 
-## Create an image with Buildah
+## Import from existing containerDisk images from kubevirt repository
```
Comment (igoihman) [inline doc/image-from-registry.md]: Import existing containerDisk images from kubevirt repository

File: doc/image-from-registry.md [hunk 2]
```diff
@@ -2,47 +2,50 @@
 The purpose of this document is to show how to create registry image containing a Virtual Machine image that can be imported into a PV.
 
 ## Prerequisites
-The VM disk image should be called 'disk.img' and placed in the root folder.
+Import from registry should be able to consume the same container images as containerDisk (https://github.com/kubevirt/kubevirt/blob/master/docs/container-register-disks.md).
```
Comment (igoihman) [inline doc/image-from-registry.md]: should be in this format:
``` [containerDisk](https://github.com/kubevirt/kubevirt/blob/master/docs/container-register-disks.md) ```

File: doc/image-from-registry.md [hunk 3]
```diff
@@ -2,47 +2,50 @@
...
+## Import from existing containerDisk images from kubevirt repository 
+For example vmidisks/fedora25:latest as described in https://github.com/kubevirt/kubevirt/blob/master/docs/container-register-disks.md
 
+## Create a container image with Buildah
 Buildah is a tool that facilitates building Open Container Initiative (OCI) container images.
-
 More information is available here: [Buildah tutorial](https://github.com/containers/buildah/blob/master/docs/tutorials/02-registries-repositories.md).
 
-Create a new container image from scratch and keep its name in a variable:
+Create a new directory /tmp/vmdisk with the following Docker file and a vm image file (ex: fedora28.qcow2)
```
Comment (igoihman) [inline doc/image-from-registry.md]: style: put `/tmp/vmdisk ` inside backticks.

File: doc/image-from-registry.md [hunk 4]
```diff
@@ -2,47 +2,50 @@
...
+cat << END > Dockerfile
+FROM kubevirt/container-disk-v1alpha
+ADD fedora28.qcow2 /disk
+END
 ```
-
-Add the VM disk to the container image and commit:
+Build and push image to a registry. 
+Note: In development environment you can push to 
+1. A cluster local cdi-docker-registry-host which hosts docker registry and is accessible within the cluster via cdi-docker-registry-host.kube-system. The registry is initialized from cluster-sync flow and is used for functional tests purposes.
```
Comment (igoihman) [inline doc/image-from-registry.md]: `cdi-docker-registry-host`, `cdi-docker-registry-host.kube-system` and `cluster-sync` inside backticks.

File: doc/image-from-registry.md [hunk 5]
```diff
@@ -2,47 +2,50 @@
...
+FROM kubevirt/container-disk-v1alpha
+ADD fedora28.qcow2 /disk
+END
 ```
-
-Add the VM disk to the container image and commit:
+Build and push image to a registry. 
+Note: In development environment you can push to 
+1. A cluster local cdi-docker-registry-host which hosts docker registry and is accessible within the cluster via cdi-docker-registry-host.kube-system. The registry is initialized from cluster-sync flow and is used for functional tests purposes. 
+2. Globally accessible registry that is used for image caching and is accessible via registry:5000 host name
```
Comment (igoihman) [inline doc/image-from-registry.md]: same for `registry:5000`

File: doc/image-from-registry.md [hunk 6]
```diff
@@ -2,47 +2,50 @@
...
-To push an image from your local Buildah container storage, check the image name, then push it using the buildah push command. Remember to identify both the local image name and a new name that includes the location (<registry>:5000, in this case):
 ```
-buildah push --tls-verify=false my-vm-img:latest <registry>:5000/my-vm-img:latest
-```
-
-## Create an image with Docker
+## Create a container image with Docker
 
-Create a Dockerfile with the following content:
+Create a Dockerfile with the following content in a new directory /tmp/vmdisk. Add an image file to the same directory (for example fedora28.qcow2)
```
Comment (igoihman) [inline doc/image-from-registry.md]: ` /tmp/vmdisk`

File: pkg/importer/dataStream.go [hunk 7]
```diff
@@ -266,6 +270,83 @@ func (d *DataStream) http() (io.ReadCloser, error) {
...
+//If such exists it creates  e Reader on it and returns it for further processing
+func (d *DataStream) registry() (io.ReadCloser, error) {
+	//generate random name for temporary directory
+	tmpDest := randTmpName(TempContainerDiskDir)
+
+	defer func(f string) {
+		os.Remove(f)
+	}(tmpDest)
+
+	imageDir := tmpDest + ContainerDiskImageDir
```
Comment (igoihman) [inline pkg/importer/dataStream.go]: better use `filepath.join`

File: pkg/importer/dataStream.go [hunk 8]
```diff
@@ -266,6 +270,83 @@ func (d *DataStream) http() (io.ReadCloser, error) {
...
+	}
+
+	//Search for file in /disk directory - if not found - failure
+	imageFile, err := getImageFileName(imageDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, fmt.Sprintf("Cannot locate image file"))
+	}
+
+	// 3. If found - Create a reader that will read this file and attach it to the dataStream
+	file, err := os.Open(imageDir + "/" + imageFile)
```
Comment (igoihman) [inline pkg/importer/dataStream.go]: `filepath.join`

File: pkg/importer/dataStream.go [hunk 9]
```diff
@@ -266,6 +270,83 @@ func (d *DataStream) http() (io.ReadCloser, error) {
...
+
+	imageDir, err := os.Open(dir)
+
+	if err != nil {
+		return "", errors.Wrapf(err, fmt.Sprintf("Cannot read from image directory"))
+	}
+
+	defer imageDir.Close()
+
+	list, err := imageDir.Readdirnames(0) // 0 to read all files and folders
```
Comment (igoihman) [inline pkg/importer/dataStream.go]: better use `fileList` - `list` is confusing.

File: doc/image-from-registry.md [hunk 10]
```diff
@@ -2,47 +2,50 @@
 The purpose of this document is to show how to create registry image containing a Virtual Machine image that can be imported into a PV.
 
 ## Prerequisites
-The VM disk image should be called 'disk.img' and placed in the root folder.
+Import from registry should be able to consume the same container images as [containerDisk](https://github.com/kubevirt/kubevirt/blob/master/docs/container-register-disks.md).
+Thus the VM disk image file to be consumed must be located under /disk directory in the container image. The file can be in any of the supported formats : qcow2, raw, archived image file. There are no any special naming contraints for the VM disk file.
```
Comment (awels) [inline doc/image-from-registry.md]: There are no special naming constraints for the VM disk.

Extra 'any' and constraints is misspelled.

**Comment by aglitke** [comment]
Looks good to me.  I think this will be ready once you address @awels comments and CI passes.

File: tests/transport_test.go [hunk 11]
```diff
@@ -123,14 +119,14 @@ var _ = Describe("Transport Tests", func() {
...
 		Entry("should not connect to http endpoint with invalid credentials", httpAuthEp, targetFile, "gopats", "bradyisthegoat", controller.SourceHTTP, false),
 		Entry("should connect to QCOW http endpoint without credentials", httpNoAuthEp, targetQCOWFile, "", "", controller.SourceHTTP, true),
 		Entry("should connect to QCOW http endpoint with credentials", httpAuthEp, targetQCOWFile, utils.AccessKeyValue, utils.SecretKeyValue, controller.SourceHTTP, true),
-		Entry("should connect to registry endpoint without credentials", registryNoAuthEp, "registry", "", "", controller.SourceRegistry, true),
-		//		Entry("should not connect to registry endpoint with invalid credentials", registryNoAuthEp, "registry", "gopats", "bradyisthegoat", controller.SourceRegistry, false),
+		Entry("should connect to registry endpoint without credentials", registryNoAuthEp, targetQCOWImage, "", "", controller.SourceRegistry, true),
+		Entry("should fail to import from registry endpoint when container image is invalid", registryNoAuthEp, targetRegistryInvalidImage, "", "", controller.SourceRegistry, false),
```
Comment (shiywang) [inline tests/transport_test.go]: I think the negtive case we want to conver is here: https://github.com/kubevirt/containerized-data-importer/pull/608/files#diff-9cbc435c7c64940bbba1090c17662234R309

should add a test about:
`should fail to import from registry endpoint when container image is not putting in  /disk directory` sth like this.

**Comment by aglitke** [comment]
@annastopel Have you figured out why this is not passing CI on openshift?

**Comment by annastopel** [comment]
@aglitke Import fails with os since cri-o containers have limited privileges. Specifically, tar fails to execute mknod and thus the entire sequence fails. I am working on a solution

**Comment by annastopel** [comment]
ci please

**Comment by annastopel** [comment]
ci test please

**Comment by annastopel** [comment]
ci test please

File: pkg/importer/dataStream.go [hunk 12]
```diff
@@ -266,6 +270,86 @@ func (d *DataStream) http() (io.ReadCloser, error) {
 	return countingReader, nil
 }
 
+const (
+	//ContainerDiskImageDir - Expected disk image location in contianer image as described in
+	//https://github.com/kubevirt/kubevirt/blob/master/docs/container-register-disks.md
+	ContainerDiskImageDir = "disk"
+	//TempContainerDiskDir - Temporary location for pulled containerImage
+	TempContainerDiskDir = "/tmp"
```
Comment (mhenriks) [inline pkg/importer/dataStream.go]: So, there is no way to use the PVC for temp space?  fakeroot did not work?

**Comment by annastopel** [comment]
I changed the implementation after our talk - now it writes to PVC.
Hopefully will push it today

File: pkg/importer/dataStream.go (hunk 12 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: Now storing on PVC in /data/tmp directory which is deleted at the end of the flow

**Comment by annastopel** [comment]
ci test please

**Comment by awels** [comment]
travis failed here: https://travis-ci.org/kubevirt/containerized-data-importer/builds/486877012#L4886

**Comment by annastopel** [comment]
> travis failed here: https://travis-ci.org/kubevirt/containerized-data-importer/builds/486877012#L4886

I am not sure this is the reason, since for https://github.com/kubevirt/containerized-data-importer/pull/628 the same error had occured and travis called it passed

**Comment by awels** [comment]
Good point, I didn't see anything remotely looking strange, so I figured that was the problem.

**Comment by annastopel** [comment]
ci test please

File: pkg/importer/dataStream.go [hunk 13]
```diff
@@ -292,9 +403,6 @@ func (d *DataStream) pollProgress(reader *util.CountingReader, idleTime, pollInt
 func CopyData(dso *DataStreamOptions) error {
 	glog.V(1).Infof("copying %q to %q...\n", dso.Endpoint, dso.Dest)
 	switch dso.Source {
-	case controller.SourceRegistry:
-		glog.V(1).Infof("using skopeo to copy from registry")
-		return image.CopyRegistryImage(dso.Endpoint, dso.Dest, dso.AccessKey, dso.SecKey)
 	default:
```
Comment (mhenriks) [inline pkg/importer/dataStream.go]: Switch statement with only default case

File: pkg/importer/dataStream.go [hunk 14]
```diff
@@ -266,6 +273,110 @@ func (d *DataStream) http() (io.ReadCloser, error) {
...
 }
 
+type registryData struct {
+	dataDir string
+	file    *os.File
+}
+
+//_regData - global variable that stores temporary data of registy flow
+// to be deleted when teh flow ends
+var _regData = registryData{"", nil}
```
Comment (mhenriks) [inline pkg/importer/dataStream.go]: Why can't this data belong to the datastream struct?

File: pkg/image/skopeo.go [hunk 15]
```diff
@@ -95,6 +95,61 @@ func CopyRegistryImage(url, dest, accessKey, secKey string) error {
...
+func CopyDirFromRegistryImage(url, dest, VMimagePath, accessKey, secKey string) error {
+	skopeoDest := "dir:" + dest + dataTmpDir
+	err := SkopeoInterface.CopyImage(url, skopeoDest, accessKey, secKey)
+	if err != nil {
+		os.RemoveAll(dest + dataTmpDir)
+		return errors.Wrap(err, "Failed to download from registry")
+	}
+	err = extractVMDiskImageDir(dest, VMimagePath)
+	if err != nil {
+		return errors.Wrap(err, "Failed to extract image layers")
```
Comment (mhenriks) [inline pkg/image/skopeo.go]: no cleaning up if we error here.

File: pkg/image/skopeo.go [hunk 16]
```diff
@@ -95,6 +95,61 @@ func CopyRegistryImage(url, dest, accessKey, secKey string) error {
...
+	}
+
+	// Clean temp folder
+	os.RemoveAll(dest + dataTmpDir)
```
Comment (mhenriks) [inline pkg/image/skopeo.go]: should probably just make this deferred.

File: pkg/importer/dataStream.go [hunk 17]
```diff
@@ -202,6 +207,8 @@ func (d *DataStream) dataStreamSelector() (err error) {
 		r, err = d.s3()
 	case "http", "https":
 		r, err = d.http()
+	case "docker", "oci":
+		r, err = d.registry()
```
Comment (mhenriks) [inline pkg/importer/dataStream.go]: unlike the other functions here like `http()`, `registry()` does a lot of stuff.  It downloads the container and extracts everything.  The bulk of the work.  Doesn't do much in `copy()` which is I believe is supposed to be the workhorse.

File: pkg/importer/dataStream.go [hunk 18]
```diff
@@ -266,6 +273,110 @@ func (d *DataStream) http() (io.ReadCloser, error) {
...
+	// 3. If found - Create a reader that will read this file and attach it to the dataStream
+	_regData.file, err = os.Open(filepath.Join(imageDir, imageFile))
+	if err != nil {
+		os.RemoveAll(_regData.dataDir)
+		return nil, errors.Wrapf(err, fmt.Sprintf("Fail to create data stream from image file"))
+	}
+
+	glog.V(1).Infof("VM disk image filename is %s", imageFile)
+
+	return RegistryDataCloser(bufio.NewReader(_regData.file)), nil
```
Comment (mhenriks) [inline pkg/importer/dataStream.go]: this seems really inefficient and could cause issues with PVC capacity.  Basically, the PVC has to be at least twice the size of the image because you are doing a copy.

File: pkg/importer/dataStream.go (hunk 14 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: This data is relevant only to import from registry flow

File: pkg/image/skopeo.go (hunk 16 above)
Comment (annastopel) [reply pkg/image/skopeo.go]: Agreed

File: pkg/importer/dataStream.go (hunk 18 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: First of all, I think we should discuss how is it better to solve temporary space issue required by all import flows.

Regarding the specific flow:
The VM disk image image we import can be in several formats:
- raw - (no redundancy see below)
- qcow2 - agree 1 redundant copy  
- qcow archived - (no redundancy see below)

- In case of raw - there is no redundant copy  - we extract the disk image file from container image and then copy it to /disk/disk.img - 
**tmpPVC** space = sizeof(containerImage)+sizeof(/disk directory in the image ~  sizeof(VM disk image file) in typical registry Image) 
**PVC space** = sizeof (raw VM disk image)

- In case of qcow2 archived - there is no redundant copy - we extract archived qcow2 file from container image and then in StreamDataToFile we unarchive it and then convert from qcow2 to raw.
**tmpPVC space** = sizeof(containerImage) + sizeof(~ archived VM disk image file)  + sizeof (unarchived VM disk image file)
**PVC space** = sizeof (raw VM disk image file)

In case of unarchive qcow2 - I agree that one copy is redundant (I am handling it) - currently 
**tmpPVC space** = sizeof(containerImage) + 2*sizeof(~  VM disk image file in qcow2 format)
**PVC space** = sizeof (raw VM disk image file)

After the fix space consumption is:
**tmpPVC space** = sizeof(containerImage) + sizeof(~  VM disk image file in qcow2 format)
**PVC space** = sizeof (raw VM disk image file)

File: pkg/image/skopeo.go (hunk 15 above)
Comment (annastopel) [reply pkg/image/skopeo.go]: Good catch!

File: pkg/importer/dataStream.go (hunk 17 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: Import form registry is different from other imports, so the handling is different. But as the rest it obtains a reader on VM image file for further processing such as (inspecting image file, its size validations and extracting it if it is archived ) Do you see a problem with such implementation?

File: pkg/importer/dataStream.go (hunk 14 above)
Comment (mhenriks) [reply pkg/importer/dataStream.go]: I still think it would be preferable to have members (even if they are unused) in the datastream.  By using global variables, we are precluding ourselves from ever doing simultaneous registry imports in the same process.

File: pkg/importer/dataStream.go (hunk 17 above)
Comment (mhenriks) [reply pkg/importer/dataStream.go]: Yeah, I think it should do as little as possible.  I don't think it has to go to the network at all, actually.  The other methods only read as little of the file as necessary to construct the appropriate readers.  The whole file is not downloaded until copy().  The registry implementation should be similar for consistency sake but also, I could imagine a case where we would want to initialize a datastream but not actually move anything until later.

Comment (mhenriks) [reply pkg/importer/dataStream.go]: Okay, I see, you have to download everything in `registry()` in order to see if the file is qcow2 so that it gets converted like an http download.  I understand now, but I'm still not sure we want to do this much work upfront

File: pkg/importer/dataStream.go (hunk 18 above)
Comment (mhenriks) [reply pkg/importer/dataStream.go]: I see redundancy in all cases.  Here's how I see it, ignoring downloading of the container layers which obliviously adds its own overhead.

Raw:

0.  Image extracted from image to `/data/disk/myimage.raw`
1.  Image COPIED from `/data/disk/myimage.raw` to `/data/disk.img`
3.  `/data/disk` directory deleted.

At step 1, there are two copies of the image on the PVC.

qcow2:

0.  Image extracted from image to `/data/disk/myimage.qcow2`
1.  Image copied to `/data/<temp dir>/myimage.qcow2`
2.  Conversion to raw creates `/data/disk.img`
3.  `/data/<temp dir>` deleted
4.  `/data/disk` deleted

At step 2, there are two copies of the qcow2 image on the pvc

qcow2 archived is similar to above except replace step 0 with:

0.  Image extracted from image to `/data/disk/myimage.qcow2.tar.gz`

File: pkg/image/skopeo.go [hunk 19]
```diff
@@ -95,6 +95,61 @@ func CopyRegistryImage(url, dest, accessKey, secKey string) error {
...
+
+	return err
+}
+
+//extractVMDiskImageDir - from each layer extracts only disk/ directory
+//Since this is the oly one that is in intereset right now
```
Comment (danielerez) [inline pkg/image/skopeo.go]: * typos: 'oly' -> 'only', 'intereset' - 'interest'
* Maybe worth keeping the signature more generalised. I.e. 'extractImageDir = func(dest, imagePath string)'.

File: pkg/importer/dataStream.go [hunk 20]
```diff
@@ -266,6 +273,110 @@ func (d *DataStream) http() (io.ReadCloser, error) {
 	return countingReader, nil
 }
 
+type registryData struct {
+	dataDir string
+	file    *os.File
+}
+
+//_regData - global variable that stores temporary data of registy flow
+// to be deleted when teh flow ends
```
Comment (danielerez) [inline pkg/importer/dataStream.go]: s/teh/the

File: pkg/importer/dataStream.go [hunk 21]
```diff
@@ -266,6 +273,110 @@ func (d *DataStream) http() (io.ReadCloser, error) {
...
+//This import source downloads specified container image from registry location
+//Then it extracts the image to a temporary location and expects an image file to be located under /disk directory
+//If such exists it creates a Reader on it and returns it for further processing
+func (d *DataStream) registry() (io.ReadCloser, error) {
+	//generate random name for temporary directory
+	_regData.dataDir = filepath.Join(d.DataDir, TempContainerDiskDir)
+
+	imageDir := filepath.Join(_regData.dataDir, ContainerDiskImageDir)
+
+	//create temporary directory if note exists to which all the data will be extracted
```
Comment (danielerez) [inline pkg/importer/dataStream.go]: note exists -> doesn't exist

File: pkg/importer/dataStream.go [hunk 22]
```diff
@@ -266,6 +273,110 @@ func (d *DataStream) http() (io.ReadCloser, error) {
...
+	}
+
+	//Search for file in /disk directory - if not found - failure
+	imageFile, err := getImageFileName(imageDir)
+	if err != nil {
+		os.RemoveAll(_regData.dataDir)
+		return nil, errors.Wrapf(err, fmt.Sprintf("Cannot locate image file"))
+	}
+
+	// 3. If found - Create a reader that will read this file and attach it to the dataStream
```
Comment (danielerez) [inline pkg/importer/dataStream.go]: add '1.' in line 326, and '2.' in line 334 :)

File: pkg/util/util.go [hunk 23]
```diff
@@ -113,6 +113,38 @@ func UnArchiveTar(reader io.Reader, destDir string, arg ...string) error {
 	return nil
 }
 
+// ExrtactSingleDirTar unarchives a tar file and streams its files
+// using the specified io.Reader to the specified destination.
+func ExrtactSingleDirTar(reader io.Reader, destDir string, extractDir string) error {
```
Comment (danielerez) [inline pkg/util/util.go]: maybe we could modify 'UnArchiveTar' function and reuse it here, instead of copy the logic. could be done later on..

File: pkg/util/util.go [hunk 24]
```diff
@@ -113,6 +113,38 @@ func UnArchiveTar(reader io.Reader, destDir string, arg ...string) error {
...
+	err = untar.Wait()
+	if err != nil {
+		glog.V(3).Infof("%s\n", string(errBuf.Bytes()))
+		return err
+	}
+
+	return nil
+}
+
+// UnArchiveSingleDirTar extracts a  extractDir from local tar file to the specified destination.
```
Comment (danielerez) [inline pkg/util/util.go]: maybe rephrase: 'extracts a specified extractDir..'

File: pkg/util/util.go [hunk 25]
```diff
@@ -113,6 +113,38 @@ func UnArchiveTar(reader io.Reader, destDir string, arg ...string) error {
...
+func UnArchiveSingleDirTar(filePath, destDir string, extractDir string) error {
```
Comment (danielerez) [inline pkg/util/util.go]: maybe call it 'UnArchiveSingleDirFromLocalTar' to make it more explicit and correlated with UnArchiveLocalTar

File: pkg/importer/dataStream.go [hunk 26]
```diff
@@ -266,6 +273,110 @@ func (d *DataStream) http() (io.ReadCloser, error) {
...
+	//copy image from registry to the temporary location
+	glog.V(1).Infof("using skopeo to copy from registry")
+	err := image.CopyDirFromRegistryImage(d.Endpoint, _regData.dataDir, ContainerDiskImageDir, d.AccessKey, d.SecKey)
+	if err != nil {
+		os.RemoveAll(_regData.dataDir)
+		return nil, errors.Wrapf(err, fmt.Sprintf("Failed ro read from registry"))
+	}
+
+	//Search for file in /disk directory - if not found - failure
+	imageFile, err := getImageFileName(imageDir)
```
Comment (mhenriks) [inline pkg/importer/dataStream.go]: Is it our intention that import from registry only supports the `kubevirt` content type?

Comment (annastopel) [reply pkg/importer/dataStream.go]: yes

File: pkg/importer/dataStream.go [hunk 27]
```diff
@@ -266,6 +273,110 @@ func (d *DataStream) http() (io.ReadCloser, error) {
...
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		return "", errors.Wrapf(err, fmt.Sprintf("image directory does not exist "))
+	}
+
+	entries, err := ioutil.ReadDir(dir)
+	if len(entries) == 0 || err != nil {
+		return "", errors.Wrapf(err, fmt.Sprintf("image file does not exist in image directory"))
+	}
+
+	fileinfo := entries[len(entries)-1]
```
Comment (mhenriks) [inline pkg/importer/dataStream.go]: Should we maybe add a little more intelligence here if there are multiple files?  Like, if there happens to be multiple files, but one is named `disk.img` use that?  Or maybe inspect content types?  Or maybe it's an error?

Comment (annastopel) [reply pkg/importer/dataStream.go]: this logic is the same as in containerDisk kubevirt/cmd/container-disk-v1alpha/entry-point.sh. Our intention is to have the same logic. we shall not assume that the disk image file has specific name or format

File: pkg/image/skopeo.go [hunk 28]
```diff
@@ -95,6 +95,61 @@ func CopyRegistryImage(url, dest, accessKey, secKey string) error {
 	return err
 }
 
+// CopyDirFromRegistryImage download image from registry with skopeo and extracts a specific directory from it
+func CopyDirFromRegistryImage(url, dest, VMimagePath, accessKey, secKey string) error {
```
[the remark this answers is not in the document]
Comment (annastopel) [reply pkg/image/skopeo.go]: agreed

File: pkg/image/skopeo.go (hunk 19 above)
Comment (annastopel) [reply pkg/image/skopeo.go]: agreed

File: pkg/importer/dataStream.go (hunk 20 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: removed

File: pkg/image/skopeo.go (hunk 16 above)
Comment (awels) [reply pkg/image/skopeo.go]: if you defer this above the error checks, then you don't need to worry about it if any of the errors happen since it will get cleaned up anyway.

File: pkg/importer/dataStream.go [hunk 29]
```diff
@@ -266,6 +273,110 @@ func (d *DataStream) http() (io.ReadCloser, error) {
...
+	io.Reader
+}
+
+func (registryDataCloser) Close() error {
+	_regData.file.Close()
+	os.RemoveAll(_regData.dataDir)
+	return nil
+}
+
+// RegistryDataCloser returns a ReadCloser with a Close method wrapping delettion
```
Comment (awels) [inline pkg/importer/dataStream.go]: // RegistryDataCloser returns a ReadCloser with a Close method wrapping deletion

File: pkg/importer/dataStream.go [hunk 30]
```diff
@@ -266,6 +273,110 @@ func (d *DataStream) http() (io.ReadCloser, error) {
...
+// of temporary data of registry flow
+func RegistryDataCloser(r io.Reader) io.ReadCloser {
+	return registryDataCloser{r}
+}
+
+const (
+	//ContainerDiskImageDir - Expected disk image location in contianer image as described in
```
Comment (awels) [inline pkg/importer/dataStream.go]: contianer -> container

File: pkg/importer/dataStream.go (hunk 13 above)
Comment (awels) [reply pkg/importer/dataStream.go]: Michael has a good point, at this point there no reason to have switch statement if we just have default and nothing else.

File: pkg/util/util.go (hunk 23 above)
Comment (awels) [reply pkg/util/util.go]: There is also a typo in the name it should be ExtractSingleDirTar

File: pkg/importer/dataStream.go (hunk 17 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: Currently leave it as is

File: pkg/importer/dataStream.go (hunk 18 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: optimized in case of qcow2 file - now it is not copied to additional tmp location, but converted to raw from the location registry flow extracted it to

**Comment by annastopel** [comment]
ci test please

File: pkg/importer/dataStream.go [hunk 31]
```diff
@@ -119,6 +130,32 @@ var rdrTypM = map[string]int{
...
+func (d *DataStream) isCopyToTmpRequired() bool {
+	//if it is unarchived qcow file that is already stored
+	//then no copy to tmp is required
+	if d.isQemu() && d.isStored() && !d.isArchived() {
+		return false
+	}
+	return true
+}
+
+func (d *DataStream) getTmpDataPath() string {
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply pkg/importer/dataStream.go]: I smell a java programmer!

File: pkg/importer/dataStream.go [hunk 32]
```diff
@@ -732,21 +910,26 @@ func (d *DataStream) copy(dest string) error {
...
 		glog.V(2).Infof("copying image file to %q", dest)
 		tmpDest := dest
-		if d.qemu {
-			// copy to tmp; qemu conversion will write to passed-in destination
-			tmpDest = randTmpName(dest)
-			glog.V(3).Infof("Copy: temp file for qcow2 conversion: %q", tmpDest)
-			defer func(f string) {
-				os.Remove(f)
-			}(tmpDest)
+		if d.isCopyToTmpRequired() {
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply pkg/importer/dataStream.go]: Please, please, please rename this function

File: pkg/importer/dataStream.go [hunk 33]
```diff
@@ -734,31 +920,51 @@ func (d *DataStream) convertQcow2ToRaw(src, dest string) error {
...
+				// copy to tmp; qemu conversion will write to passed-in destination
+				tmpDest = randTmpName(dest)
+				glog.V(3).Infof("Copy: temp file for qcow2 conversion: %q", tmpDest)
+				defer func(f string) {
+					os.Remove(f)
+				}(tmpDest)
+			}
+			if d.isMovble() {
+				//before moving has to close all readers
+				closeReaders(d.Readers)
```
Comment (mhenriks) [inline pkg/importer/dataStream.go]: Shouldn't be necessary

File: pkg/importer/dataStream.go [hunk 34]
```diff
@@ -734,31 +920,51 @@ func (d *DataStream) convertQcow2ToRaw(src, dest string) error {
...
+				err = MoveFile(d.getTmpDataPath(), tmpDest)
+				if err != nil {
+					//Failback to copy since move failed
```
Comment (mhenriks) [inline pkg/importer/dataStream.go]: I what situations would this code succeed if a move failed?

File: pkg/importer/dataStream.go [hunk 35]
```diff
@@ -734,31 +920,51 @@ func (d *DataStream) convertQcow2ToRaw(src, dest string) error {
...
+					file, fileerr := os.Open(d.getTmpDataPath())
```
Comment (mhenriks) [inline pkg/importer/dataStream.go]: You wouldn't have to bother reopening this file if you didn't close readers above

File: pkg/importer/dataStream.go (hunk 33 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: At this stage there are Readers on the file,either Rename will fail or the file handle sill become stale and Close in dataStream.Close() will throw an error

File: pkg/importer/dataStream.go (hunk 34 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: If source and dest belong to different devices/partitions (tmpPVC vs PVC) Rename will fail with the following error "invalid cross-device link". Copy however, does succeed, since it will actually create a new file on the destination file system

File: pkg/importer/dataStream.go (hunk 35 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: Yes, but then again - Rename will fail due to opened file handles

File: pkg/importer/dataStream.go (hunk 32 above)
[the remark this answers is not in the document]
Comment (annastopel) [reply pkg/importer/dataStream.go]: Done

File: pkg/importer/dataStream.go (hunk 33 above)
Comment (mhenriks) [reply pkg/importer/dataStream.go]: That is very surprising to me.  Renaming an open file is totally fine at the POSIX level so I wonder where/how this error is triggered.

File: pkg/importer/dataStream.go (hunk 34 above)
Comment (mhenriks) [reply pkg/importer/dataStream.go]: oh, got it

File: pkg/importer/dataStream.go [hunk 36]
```diff
@@ -734,31 +920,51 @@ func (d *DataStream) convertQcow2ToRaw(src, dest string) error {
...
+					if fileerr != nil {
+						glog.Errorf(fileerr.Error(), "Failed to open image file ", d.getTmpDataPath)
+						return errors.WithMessage(fileerr, fmt.Sprintf("failed to open image file when failing back to copy"))
+					}
+					d.appendReader(rdrFile, bufio.NewReader(file))
+					err = StreamDataToFile(d.topReader(), tmpDest)
```
Comment (mhenriks) [inline pkg/importer/dataStream.go]: I think it would be cleaner to just do `io.Copy` or `ioutil.WriteFile` here.

File: pkg/importer/dataStream.go (hunk 33 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: You are absolutely right!
 i got confused with src opened file vs dest opened file. Pushing a fix for that

File: pkg/importer/dataStream.go (hunk 35 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: You are right! pushing the fix

**Comment by annastopel** [comment]
ci test please

**Comment by annastopel** [comment]
ci test please

File: pkg/importer/dataStream.go [hunk 37]
```diff
@@ -121,6 +132,37 @@ var rdrTypM = map[string]int{
...
+func (d *DataStream) isQemuStored() bool {
+	//if it is unarchived qcow file that is already stored
+	//then no copy to tmp is required
+	if d.isQemu() && d.isStored() && !d.isArchived() {
+		return true
+	}
+	return false
+}
+
+func (d *DataStream) isMovble() bool {
```
Comment (awels) [inline pkg/importer/dataStream.go]: isMoveable()

File: pkg/importer/dataStream.go [hunk 38]
```diff
@@ -269,6 +328,128 @@ func (d *DataStream) http() (io.ReadCloser, error) {
...
+	}
+
+	filename := fileinfo.Name()
+
+	if len(strings.TrimSpace(filename)) == 0 {
+		glog.Errorf("image file does not exist in image directory - file has no name ")
+		return "", errors.Errorf(fmt.Sprintf("image file does not exist in image directory"))
+	}
+
+	glog.V(1).Infof("VM disk image filename is %s", filename)
```
Comment (awels) [inline pkg/importer/dataStream.go]: I don't think this is a verbosity 1 level message, its more debugging, so a 3 is more appropriate IMO.

File: pkg/importer/dataStream.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (awels) [reply pkg/importer/dataStream.go]: I am going to have to agree with Michael here, the extra functions that simply return a field value, have little value. isStored actually does a calculation so I have no issues with that one, but isQemu, and isArchived don't give you any value. d.qemu vs d.isQemu() and d.archived vs d.isArchived() aren't that helpful.

[the remark this answers is not in the document]
Comment (annastopel) [reply pkg/importer/dataStream.go]: done

File: pkg/importer/dataStream.go (hunk 37 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: done

File: pkg/importer/dataStream.go (hunk 38 above)
Comment (annastopel) [reply pkg/importer/dataStream.go]: done

**Comment by annastopel** [comment]
ci test please

**Comment by annastopel** [comment]
ci test please
